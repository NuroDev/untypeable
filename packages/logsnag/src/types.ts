export type * from "./log/log.types";
export type * from "./insight/insight.types";
