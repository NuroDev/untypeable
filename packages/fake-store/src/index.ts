import { initUntypeable } from "untypeable";

import type {
  Cart,
  CartParams,
  Carts,
  CartsParams,
  CreateCartParams,
  DeleteCartParams,
  UpdateCartParams,
  UserCartsParams,
} from "./cart/cart.types";
import type {
  Product,
  ProductsParams,
  Products,
  CreateProductParams,
  UpdateProductParams,
  ProductParams,
  ProductsCategories,
  ProductsCategoriesParams,
  ProductsCategoryParams,
  ProductsCategory,
  DeleteProductParams,
  UpdateProduct,
  CreateProduct,
} from "./products/products.types";

const u = initUntypeable().args<
  "GET" | "POST" | "PUT" | "PATCH" | "DELETE",
  string
>();

const cartRouter = u.router({
  DELETE: {
    /** Delete a cart */
    "/carts/:id": u.input<DeleteCartParams>().output<Cart>(),
  },
  GET: {
    /** Get all carts */
    "/carts": u.input<CartsParams>().output<Carts>(),
    /** Get a single cart */
    "/carts/:id": u.input<CartParams>().output<Cart>(),
    /** Get user carts */
    "/carts/user/:userId": u.input<UserCartsParams>().output<UserCartsParams>(),
  },
  PATCH: {
    /** Update a product */
    "/carts/:id": u.input<UpdateCartParams>().output<Cart>(),
  },
  POST: {
    /** Add a new product */
    "/carts": u.input<CreateCartParams>().output<Cart>(),
  },
  PUT: {
    /** Update a product */
    "/carts/:id": u.input<UpdateCartParams>().output<Cart>(),
  },
});

const loginRouter = u.router({
  DELETE: {},
  GET: {},
  PATCH: {},
  POST: {},
  PUT: {},
});

const productRouter = u.router({
  DELETE: {
    /** Delete a product */
    "/products/:id": u.input<DeleteProductParams>().output<Product>(),
  },
  GET: {
    /** Get all products */
    "/products": u.input<ProductsParams>().output<Products>(),
    /** Get a single product */
    "/products/:id": u.input<ProductParams>().output<Product>(),
    /** Get all categories */
    "/products/categories": u
      .input<ProductsCategoriesParams>()
      .output<ProductsCategories>(),
    /** Get products in a specific category */
    "/products/category/:category": u
      .input<ProductsCategoryParams>()
      .output<ProductsCategory>(),
  },
  PATCH: {
    /** Update a product */
    "/products/:id": u.input<UpdateProductParams>().output<UpdateProduct>(),
  },
  POST: {
    /** Add a new product */
    "/products": u.input<CreateProductParams>().output<CreateProduct>(),
  },
  PUT: {
    /** Update a product */
    "/products/:id": u.input<UpdateProductParams>().output<UpdateProduct>(),
  },
});

const userRouter = u.router({
  DELETE: {},
  GET: {},
  PATCH: {},
  POST: {},
  PUT: {},
});

const router = u
  .router({})
  .merge(productRouter)
  .merge(cartRouter)
  .merge(userRouter)
  .merge(loginRouter);

export type FakeStoreRouter = typeof router;
