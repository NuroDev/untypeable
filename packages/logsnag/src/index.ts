import { initUntypeable } from "untypeable";

import type { Insight } from "./insight.types";
import type { Log } from "./log.types";

const u = initUntypeable();

const router = u.router({
  "/insight": u.input<Insight>().output<Insight>(),
  "/log": u.input<Log>().output<Log>(),
});

export type LogSnagRouter = typeof router;
