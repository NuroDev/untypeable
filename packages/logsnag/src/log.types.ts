import type { z } from "zod";

import type { LogSchema } from "./log.validators";

export type Log = z.infer<typeof LogSchema>;
