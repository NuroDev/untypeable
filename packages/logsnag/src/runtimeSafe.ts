import { initUntypeable } from "untypeable";

import { InsightSchema } from "./insight/insight.validators";
import { LogSchema } from "./log/log.validators";

const u = initUntypeable();

export const logSnagSafeRouter = u.router({
  "/insight": u.input(InsightSchema).output(InsightSchema),
  "/log": u.input(LogSchema).output(LogSchema),
});
