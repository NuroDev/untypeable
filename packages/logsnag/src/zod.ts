export * from "./log.validators";
export * from "./insight.validators";
