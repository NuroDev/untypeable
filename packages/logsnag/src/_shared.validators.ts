import { z } from "zod";

export const SharedLogSnagParamsSchema = z.object({
  apiVersion: z.enum(["v1"]).optional(),
  /**
   * Project name
   */
  project: z.string(),
});
