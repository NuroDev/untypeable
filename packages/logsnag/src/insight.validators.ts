import { z } from "zod";

import { SharedLogSnagParamsSchema } from "./_shared.validators";

export const InsightSchema = SharedLogSnagParamsSchema.extend({
  /**
   * Event icon (emoji)
   * must be a single emoji
   * example: "ðŸ‘¨"
   */
  icon: z.string().optional(),
  /**
   * Insight title
   * example: "User Count"
   */
  title: z.string(),
  /**
   * Insight value
   * example: 100
   */
  value: z.string().or(z.boolean()).or(z.number()),
});
