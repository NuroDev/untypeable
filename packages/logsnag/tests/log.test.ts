import { describe, it, expect } from "vitest";

import { LogSchema } from "../src/zod";
import { useTestClient } from "./_client";

describe.concurrent("LogSnag - Log", () => {
  const client = useTestClient();

  it("POST - /log", async () => {
    const log = await client("/log", {
      project: "nuro",
      channel: "untyped",
      event: "Test",
      description: "A basic test event from the `@untyped/logsnag` package",
      icon: "ðŸ‘‹",
      notify: false,
    });

    expect(log).toBeDefined();
    expect(log).toMatchSnapshot({
      project: "nuro",
      channel: "untyped",
      event: "Test",
      description: "A basic test event from the `@untyped/logsnag` package",
      icon: "ðŸ‘‹",
      notify: false,
      parser: "text",
    });

    expect(LogSchema.safeParse(log).success).toBe(true);
  });
});
