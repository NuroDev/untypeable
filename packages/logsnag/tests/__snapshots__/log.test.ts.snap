// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogSnag - Log > POST - /log 1`] = `
{
  "channel": "untyped",
  "description": "A basic test event from the \`@untyped/logsnag\` package",
  "event": "Test",
  "icon": "ðŸ‘‹",
  "notify": false,
  "parser": "text",
  "project": "nuro",
}
`;
